packages/ens-utils/src/ethregistrar.ts

```diff
@@ -155,3 +179,284 @@ export function buildNFTRefFromENSName(
+    ),
+    timestamp: temporaryPremiumEndTimestamp,
+  });
+  return {
+    relativeTimestamp: formatTimestampAsDistanceToNow(
+      temporaryPremiumEndTimestamp,
+    ),
+    timestamp: temporaryPremiumEndTimestamp,
+  };
+};
+
+/**
+ * At the moment a .eth name expires, this recently released temporary premium is added to its price.
+ * NOTE: The actual recently released temporary premium added subtracts `PREMIUM_OFFSET`.
+ */
+export const PREMIUM_START_PRICE: Price = {
+  value: 10000000000n /* $100,000,000.00 (100 million USD) */,
+  currency: Currency.Usd,
+};
+
+/**
+ * The recently released temporary premium drops exponentially by 50% each day.
+ */
+const PREMIUM_DECAY = 0.5;
+
+/**
+ * Goal:
+ *  The temporary premium should drop to $0.00 after exactly `PREMIUM_DAYS` days have passed.
+ *
+ * Challenge:
+ *  If we decay `PREMIUM_START` by a rate of `PREMIUM_DECAY` each day over the course of
+ *  `PREMIUM_DAYS` days we don't get $0.00 USD. Instead, we get this `PREMIUM_OFFSET` value
+ *  ($47.68 USD).
+ *
+ * Solution:
+ *  Subtract this value from the decayed temporary premium to get the actual temporary premium.
+ */
+export const PREMIUM_OFFSET = approxScalePrice(
+  PREMIUM_START_PRICE,
+  PREMIUM_DECAY ** Number(TEMPORARY_PREMIUM_DAYS),
+);
+
+export function temporaryPremiumPriceAtTimestamp(
+  atTimestamp: Timestamp,
+  expirationTimestamp: Timestamp,
+): Price {
+  const releasedTimestamp = addSeconds(expirationTimestamp, GRACE_PERIOD);
+  const secondsSinceRelease = atTimestamp.time - releasedTimestamp.time;
+  if (secondsSinceRelease < 0) {
+    /* if as of the moment of `atTimestamp` a name hasn't expired yet then there is no temporaryPremium */
+    return {
+      value: 0n,
+      currency: Currency.Usd,
+    };
+  }
+
+  const fractionalDaysSinceRelease =
+    Number(secondsSinceRelease) / Number(SECONDS_PER_DAY.seconds);
+
+  const decayFactor = PREMIUM_DECAY ** fractionalDaysSinceRelease;
+
+  const decayedPrice = approxScalePrice(PREMIUM_START_PRICE, decayFactor);
+  const offsetDecayedPrice = subtractPrices(decayedPrice, PREMIUM_OFFSET);
+
+  /* the temporary premium can never be less than $0.00 */
+  if (offsetDecayedPrice.value < 0n) {
+    return {
+      value: 0n,
+      currency: Currency.Usd,
+    };
+  }
+
+  return offsetDecayedPrice;
+}
+
+/*
+  This is an "internal" helper function only. It can't be directly used anywhere else because
+  it is too easy to accidently not include the registration object when it should be passed.
+  Three different functions are created right below this one, which are the ones that are
+  safe to be used across the platform, and are then, the ones being exported.
+*/
+const AvailableNamePriceUSD = (
+  parsedName: DomainName,
+  registerForYears = DEFAULT_REGISTRATION_YEARS,
+  registration: Registration | null = null,
+  additionalFee: Price | null = null,
+): Price | null => {
+  if (!parsedName.normalizedName) return null;
+
+  const defaultPrice: Readonly<Price> = {
```

Comment:
Suggest extracting the price constants for 3, 4, and 5 character names out of this function and into the overall file. There's no need for us to rebuild them all each time this function is called.

Reply:
Should this be exported?

Reply:
No. Functions that determine the appropriate price can be exported, but this constant should remain internal.